bgpd/bgp_conditional_adv.c

```diff
@@ -0,0 +1,451 @@
+								 : RMAP_NOOP;
+
+			if ((ret == RMAP_NOOP) && (prev_ret == RMAP_NOOP))
+				continue;
+
+			/* Derive conditional advertisement status from
+			 * condition and return value of condition-map
+			 * validation.
+			 */
+			if (adv_conditional == NLRI)
+				filter->advmap.status = true;
+			else if (adv_conditional == WITHDRAW)
+				filter->advmap.status = false;
+			else {
+				/* no change in advertise status. So, only
+				 * previously withdrawn routes will be
+				 * advertised if needed.
+				 */
+			}
+
+			/* Send regular update as per the existing policy.
+			 * There is a change in route-map, match-rule, ACLs,
+			 * or route-map filter configuration on the same peer.
+			 */
+			if (peer->advmap_info[afi][safi].config_change) {
+				paf = peer_af_find(peer, afi, safi);
+				if (paf) {
+					update_subgroup_split_peer(paf, NULL);
+					subgrp = paf->subgroup;
+					if (subgrp && subgrp->update_group)
+						subgroup_announce_table(
+							paf->subgroup, NULL);
+				}
+				peer->advmap_info[afi][safi].config_change =
+					false;
+			}
+
+			/* Send update as per the conditional advertisement */
+			if (adv_conditional) {
+				route_advertised = bgp_conditional_adv_routes(
+					peer, afi, safi, table,
+					filter->advmap.amap,
+					filter->advmap.status);
+
+				/* amap_prev_status is only to check whether we
+				 * have announced any routes(advertise/withdraw)
+				 * or not. filter->advmap.status will have the
+				 * actual filter status
+				 */
+				peer->advmap_info[afi][safi].amap_prev_status =
+					route_advertised ? RMAP_MATCH
+							 : RMAP_NOOP;
+			}
+		}
+	}
+	return 0;
+}
+
+void bgp_conditional_adv_enable(struct peer *peer, afi_t afi, safi_t safi)
+{
+	struct bgp *bgp = peer->bgp;
+
+	assert(bgp);
+
+	/* This flag is used to monitor conditional routes status in BGP table,
+	 * and advertise/withdraw routes only when there is a change in BGP
+	 * table w.r.t conditional routes
+	 */
+	peer->advmap_info[afi][safi].amap_prev_status = RMAP_NOOP;
+	peer->advmap_info[afi][safi].cmap_prev_status = RMAP_NOOP;
+	peer->advmap_info[afi][safi].config_change = true;
+
+	/* advertise-map is already configured on atleast one of its
+	 * neighbors (AFI/SAFI). So just increment the counter.
+	 */
+	if (++bgp->condition_filter_count > 1)
+		return;
+
+	/* Register for conditional routes polling timer */
+	thread_add_timer(bm->master, bgp_conditional_adv_timer, bgp,
+			 CONDITIONAL_ROUTES_POLL_TIME, &bgp->t_condition_check);
+}
+
+void bgp_conditional_adv_disable(struct peer *peer, afi_t afi, safi_t safi)
+{
+	struct bgp *bgp = peer->bgp;
+
+	assert(bgp);
+
+	/* advertise-map is not configured on any of its neighbors or
+	 * it is configured on more than one neighbor(AFI/SAFI).
+	 * So there's nothing to do except decrementing the counter.
+	 */
+	if (--bgp->condition_filter_count != 0)
+		return;
+
+	/* Last filter removed. So cancel conditional routes polling thread. */
+	THREAD_OFF(bgp->t_condition_check);
```

Comment:
double-checking: Is this a global thread or should be per-peer?

Reply:
its global thread. not per peer. 
when advertise-map is configured on first peer it will be created and deleted when advertise-map on last peer is removed.

Reply:
and what happens if I have 100 peers each with its own advertise-map?

Reply:
thread will be deleted/OFF only after removing the advertise-map config on 100th peer.

Reply:
I'm not sure if this is the right way. Does it mean that with N peers it would take N*CONDITIONAL_ROUTES_POLL_TIME to complete?

Reply:
No. Always it will be 60 seconds, irrespective of the number of advertise-maps configured per BGP process.
when scanning process is called we will check if peer[afi][safi] has advertise-map configurations and process the pending event on each peer (AFI/SAFI).